cmd/cluster-agent-cloudfoundry/app/app.go

```diff
@@ -0,0 +1,269 @@
+	if !config.Datadog.IsSet("api_key") {
+		log.Critical("no API key configured, exiting")
+		return nil
+	}
+
+	// Setup healthcheck port
+	var healthPort = config.Datadog.GetInt("health_port")
+	if healthPort > 0 {
+		err := healthprobe.Serve(mainCtx, healthPort)
+		if err != nil {
+			return log.Errorf("Error starting health port, exiting: %v", err)
+		}
+		log.Debugf("Health check listening on port %d", healthPort)
+	}
+
+	// get hostname
+	hostname, err := util.GetHostname()
+	if err != nil {
+		return log.Errorf("Error while getting hostname, exiting: %v", err)
+	}
+	log.Infof("Hostname is: %s", hostname)
+
+	// setup the forwarder
+	keysPerDomain, err := config.GetMultipleEndpoints()
+	if err != nil {
+		log.Error("Misconfiguration of agent endpoints: ", err)
+	}
+	f := forwarder.NewDefaultForwarder(keysPerDomain)
+	f.Start()
+	s := serializer.NewSerializer(f)
+
+	aggregatorInstance := aggregator.InitAggregator(s, nil, hostname, "cluster_agent")
+	aggregatorInstance.AddAgentStartupTelemetry(fmt.Sprintf("%s - Datadog Cluster Agent", version.AgentVersion))
+
+	log.Infof("Datadog Cluster Agent is now running.")
+
+	// Setup a channel to catch OS signals
+	signalCh := make(chan os.Signal, 1)
+	signal.Notify(signalCh, os.Interrupt, syscall.SIGTERM)
+
+	// initialize BBS Cache before starting provider/listener
+	err = initializeBBSCache()
+	if err != nil {
+		return err
+	}
+
+	// create and setup the Autoconfig instance
+	common.SetupAutoConfig(config.Datadog.GetString("confd_path"))
+	// start the autoconfig, this will immediately run any configured check
+	common.StartAutoConfig()
+
+	var clusterCheckHandler *clusterchecks.Handler
+	if config.Datadog.GetBool("cluster_checks.enabled") {
+		// Start the cluster check Autodiscovery
+		clusterCheckHandler, err = setupClusterCheck(mainCtx)
+		if err != nil {
+			log.Errorf("Error while setting up cluster check Autodiscovery %v", err)
+		}
+	} else {
+		log.Debug("Cluster check Autodiscovery disabled")
+	}
+
+	// Start the cmd HTTPS server
+	// We always need to start it, even with nil clusterCheckHandler
+	// as it's also used to perform the agent commands (e.g. agent status)
+	sc := clusteragent.ServerContext{
+		ClusterCheckHandler: clusterCheckHandler,
+	}
+	if err = api.StartServer(sc); err != nil {
+		return log.Errorf("Error while starting agent API, exiting: %v", err)
+	}
+
+	// Block here until we receive the interrupt signal
+	<-signalCh
+
+	// retrieve the agent health before stopping the components
+	// GetStatusNonBlocking has a 100ms timeout to avoid blocking
+	health, err := health.GetStatusNonBlocking()
+	if err != nil {
+		log.Warnf("Cluster Agent health unknown: %s", err)
+	} else if len(health.Unhealthy) > 0 {
+		log.Warnf("Some components were unhealthy: %v", health.Unhealthy)
+	}
+
+	// Cancel the main context to stop components
+	mainCtxCancel()
+
+	if stopCh != nil {
+		close(stopCh)
+	}
+
+	log.Info("See ya!")
+	log.Flush()
+	return nil
+}
+
+func initializeBBSCache() error {
```

Comment:
I believe this function should only init the bbs client (with a retrier) and make sure the agent can connect - without configuring auto-discovery. IMO all auto-discovery config must be handled in `common.SetupAutoConfig`.

Reply:
So maybe I'm misunderstanding something here, but my understanding is that getting `config.ConfigurationProviders` to initialize the BBS cache doesn't actually initialize the provider. This is just a structure with the config fields and doesn't actually *do* anything. Please correct me if I'm wrong.

Reply:
Here is the current pattern used to register AD components:
[common.SetupAutoConfig](https://github.com/DataDog/datadog-agent/blob/1c76b8381a195a0b0f629011a6225e936fe1d37a/cmd/agent/common/autoconfig.go#L24) is used to configure auto-discovery by picking the right providers and listeners (in our case cloud foundry provider and listener + the file provider, to be set in the config file of the agent) that will be used by the main auto-discovery logic to generate services and configs.
The provider and listener should init themselves once called by auto-discovery.

Reply:
Done in https://github.com/DataDog/datadog-agent/pull/4935/commits/cc8a3e91a9bb59f4f1daa60fdd3cf817288dcfe1